core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -844,11 +844,11 @@ protected void doRun() {
                 // we never execute replication operation locally as primary operation has already completed locally
                 // hence, we ignore any local shard for replication
                 if (nodes.localNodeId().equals(shard.currentNodeId()) == false) {
-                    performOnReplica(shard, shard.currentNodeId());
+                    performOnReplica(shard);
                 }
                 // send operation to relocating shard
                 if (shard.relocating()) {
-                    performOnReplica(shard, shard.relocatingNodeId());
+                    performOnReplica(shard.buildTargetRelocatingShard());
```

Comment:
I'm fine with this for this fix (to keep it small), but it's a shame we create these objects. As a followup I think we should change IndexShardRoutingTable#activeShards to include relocation targets (as allInitializingShards does). Then we can use that list which is cached and we also won't need the relocation check,